### DISCUSSION TITLE
Adding a feature that will execute a function when the bot goes down via an error such as Keyboard Interruption?

### DISCUSSION THREAD

**Discussion by Sheppsu** [Rapptz/discord.py#5957] [Ideas] [answered]
I went into the source code and tried doing this myself to find some nice results. I assume it would not work when something happens such as whatever's running the bot gets forcefully closed. However, it worked with KeyboardInterruption and I wonder if anyone else would find this useful in their bots. 
If anyone is wondering what I did to the code (client.py > Client class > run function):
```Python
        death_func = kwargs.pop('death', None)

        async def runner():
            try:
                await self.start(*args, **kwargs)
            finally:
                if death_func is not None:
                    await death_func()
                if not self.is_closed():
                    await self.close()
```
The only changes are with the death_func variable, and the two lines that involve executing it.

**Comment by Rapptz** [comment, marked as answer]
You don't need to modify any source to do this. Just use standard Python functions. Your best bet is to do this before the bot is closed, so something like this:

```py
from discord.ext import commands

class MyBot(commands.Bot):
    async def my_finalizer(self):
        # do whatever here
        pass

    async def close(self):
        await self.my_finalizer()
        await super().close()
```
